Source/DafnyCore/DafnyPrelude.bpl

```diff
@@ -33,35 +34,35 @@ const unique TORDINAL  : Ty uses {
   axiom Tag(TORDINAL) == TagORDINAL;
 }
 // See for which axioms we can make use of the trigger to determine the connection.
-function TBitvector(int) : Ty uses {
-  axiom (forall w: int :: { TBitvector(w) } Inv0_TBitvector(TBitvector(w)) == w);
-}
-function TSet(Ty)      : Ty uses {
-  axiom (forall t: Ty :: { TSet(t) } Inv0_TSet(TSet(t)) == t);
-  axiom (forall t: Ty    :: { TSet(t) }      Tag(TSet(t))      == TagSet);
-}
-function TISet(Ty)     : Ty uses {
-  axiom (forall t: Ty :: { TISet(t) } Inv0_TISet(TISet(t)) == t);
-  axiom (forall t: Ty    :: { TISet(t) }     Tag(TISet(t))     == TagISet);
-}
-function TMultiSet(Ty) : Ty uses {
-  axiom (forall t: Ty :: { TMultiSet(t) } Inv0_TMultiSet(TMultiSet(t)) == t);
-  axiom (forall t: Ty    :: { TMultiSet(t) } Tag(TMultiSet(t)) == TagMultiSet);
-}
-function TSeq(Ty)      : Ty uses {
-  axiom (forall t: Ty :: { TSeq(t) } Inv0_TSeq(TSeq(t)) == t);
-  axiom (forall t: Ty    :: { TSeq(t) }      Tag(TSeq(t))      == TagSeq);
-}
-function TMap(Ty, Ty)  : Ty uses {
-  axiom (forall t, u: Ty :: { TMap(t,u) } Inv0_TMap(TMap(t,u)) == t);
-  axiom (forall t, u: Ty :: { TMap(t,u) } Inv1_TMap(TMap(t,u)) == u);
-  axiom (forall t, u: Ty :: { TMap(t,u) }    Tag(TMap(t,u))    == TagMap);
-}
-function TIMap(Ty, Ty) : Ty uses {
-  axiom (forall t, u: Ty :: { TIMap(t,u) } Inv0_TIMap(TIMap(t,u)) == t);
-  axiom (forall t, u: Ty :: { TIMap(t,u) } Inv1_TIMap(TIMap(t,u)) == u);
-  axiom (forall t, u: Ty :: { TIMap(t,u) }   Tag(TIMap(t,u))   == TagIMap);
-}
+function TBitvector(int) : Ty;
```

Comment:
We are already depending on the trigger inference of Boogie when triggers have multiple function / constant symbols, so something changing about trigger inference would already break the prelude. Moving axioms with only a single term out of their `uses` clauses would only amplify the brokenness :-) About the documentation aspect, don't triggers provide exactly the same information?

I think moving to `used_by` (or an alternative keyword without underscores), and disabling trigger inference completely would be the better approach, eventually. This takes away the possibility of misusing `uses` clauses (which would weaken pruning), and would be more robust as a result of not relying on automatic inference in Boogie.

Reply:
> Overall, this PR make a remarkable improvement to the mean and standard deviation of both time and resource count on a large set of benchmarks.

I think it would be good to understand where the differences in pruning are.

> I don't know how much is attributable to this particular change, but it's conceivable that moving these out of the uses clause could lead to more pruning.

How do you conceive that? I think trigger based inference would infer the same dependencies.

> About the documentation aspect, don't triggers provide exactly the same information?

I don't think so, because triggers are nested inside axioms, and the axiom may be placed in a location distant from the owning declaration. I think the `uses` clause provides the clearest way to say that an axiom is only relevant for this declaration. Note that Boogie also has the concept of a `DefinitionAxiom` which is a single axiom that serves to define the function, with the syntax `function A(<parameters>): ReturnType { <definitionAxiom> }`. I think this serves a similar purpose as the `uses` clause but has the downside that it supports one axiom at most.

I would argue we should keep the `uses` clauses for now even if we add `used_by`, because I think >90% of axiom cases can be covered by the more documenting `uses`.

Reply:
> I think it would be good to understand where the differences in pruning are.

The benefit arises in axioms where (a) in the following sentence holds, but (b) doesn't.

- "Do not prune if (a) s is reachable or (b) all symbols in one of its triggers are reachable."

By moving such axioms out of their `uses` clauses, we are strengthening it by removing the disjunction.

- "Do not prune if (b) all symbols in one of its triggers are reachable."

> How do you conceive that? I think trigger based inference would infer the same dependencies.

You're right, in the axioms this comment is referring to, both (a) and (b) would hold at the same time, because the axioms only have a single function symbol. In this case it doesn't matter besides the documentation aspect.

> Compared to trigger, I think the uses clause provides a clearer way to say that an axiom is only relevant for this declaration.

I think there are other ways to document such relations, such as using comments and bundling up related functions and axioms together, as done in the axiomatizations. Consider the example you gave, it currently makes no difference for pruning whether this axiom is inside a `uses` clause or not, so let's put it inside the `uses` clause for`$Box`:

```
function $Box(x: T) uses {
  axiom (forall<T> x : T :: { $Box(x) } $Unbox($Box(x)) == x);
}
```
Now let's update the trigger as you pointed out:

> Btw, should the trigger of this be updated to Unbox(Box(x)) ?

```
function $Box(x: T) uses {
  axiom (forall<T> x : T :: { $Unbox($Box(x)) } $Unbox($Box(x)) == x);
}
```
Ignoring that a `$Box`  might never appear without `$Unbox` in a program in practice, this will weaken pruning if `$Unbox` is not reachable: (a) holds and (b) doesn't. Having the axiom outside of any `uses` clauses would minimize the risk of this happening.

(Btw I am not sure if `$Unbox` should be added there. `$Box` would not be injective without this axiom. In any case, I think adjusting triggers is beyond the scope of this PR - this PR only aims to make unused axioms prunable.)

`Maybe we should replace uses clauses by definition axioms where possible`

I think this might make sense, does this happen when one provides a definition for a function and not provide the `:inline` attribute? One potential risk here is if `:inline` attribute is later added to such functions, which would break axioms where this function appears in a trigger.

> You can incorrectly use used_by just as well, can you not? I think one could argue for doing as much as possible based on the triggers so there's less possibility of specifying too many dependencies.

I agree, and I am starting to think it might be best if edge inference from triggers is preserved. The only correct way to use `used_by` might be to add exactly one tuple for each trigger of an axiom, which the current edge inference is already doing. Decoupling the semantics would make it possible to strengthen pruning, but am having difficulty seeing any cases where this would be useful. 

One potential application might be unquantified axioms where edge inference does not apply, and `used_by` allows for finer control over pruning logic than `uses` does, but I didn't see any cases in the prelude where this would benefit.